drivers/include/soft_uart.h

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (C) 2020 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    drivers_soft_uart Software UART
+ * @ingroup     drivers_soft_periph
+ * @brief       Software implemented UART
+ *
+ * This module provides a software implemented Universal Asynchronous Receiver Transmitter.
+ * It is intended to be used in situation where hardware UART is not available.
+ * The signatures of the functions are similar to the functions declared in uart.h
+ *
+ * Currently sending and receiving is not possible at the same time, so loopback operation
+ * is not possible.
+ *
+ * @{
+ *
+ * @file
+ * @brief       Software UART port descriptor definition
+ *
+ * @author      Benjamin Valentin <benjjamin.valentin@ml-pa.com>
+ */
+
+#ifndef SOFT_UART_H
+#define SOFT_UART_H
+
+#include "periph/gpio.h"
+#include "periph/uart.h"
+#include "periph/timer.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief invert the level of the TX signal
+ */
+#define SOFT_UART_FLAG_INVERT_TX    0x1
+
+/**
+ * @brief invert the level of the RX signal
+ */
+#define SOFT_UART_FLAG_INVERT_RX    0x2
+
+/**
+ * @brief Software UART port descriptor
+ */
+typedef struct {
+    gpio_t rx_pin;          /**< RX pin */
+    gpio_t tx_pin;          /**< TX pin */
+    tim_t rx_timer;         /**< Hardware timer used for RX */
+    tim_t tx_timer;         /**< Hardware timer used for TX */
```

Comment:
A single time with two channels should also work. Especially as the code internally assumes both timers are running on the same frequency, which might not be possible with distinct timers.

On MCUs that only have a single timer channel, multiplexing should not be too hard for this trivial case. I think the following approach should work:

- The RX code will set the timeout only, if no TX is ongoing.
- The RX timeout is also stored in RAM
- The TX code will just always set the timeout, overwriting any RX timeouts if present
- If an outgoing byte is fully send, the TX code will check if an RX timeout was missed (and handle it then), an RX timeout is pending (and set the timer then), or no action is required
- The timeout ISR will check if TX is ongoing, in which case it always is handling a TX timeout. Otherwise an RX timeout is handled
- The GPIO ISR of the RX pin needs to take into account that it might be triggered by the start bit of the next byte, while the previous byte is not yet passed to upper layer (due to delayed RX timeout).

Reply:
Not sure how two channels would help here, there still is only one counter.

The timers only run if a byte is received / sent.
And the TX timer will wrap after each bit.

Granted, this doesn't really work in full-duplex mode anyway as the interrupts will cause too much of an overhead to keep the timing accurate…

Reply:
> And the TX timer will wrap after each bit.

OK, I see. Using two timers still seems like it would prevent its use on some boards that might want it. E.g. when porting a new platform to RIOT, UART currently is one of the high prio targets just for stdio. This might change with this PR being merged. But if a platform would only have two timers in total, using both xtimer/ztimer and the soft UART would not work like this.

> Granted, this doesn't really work in full-duplex mode anyway

Maybe the RX implementation could be changed to also periodically sample the input instead. In the timer ISR one would first send the output bit (if any) and then sample the RX bit. The timer would be stopped once both RX and TX are completed. That should both fix the issue with full duplex and only require a single timer.

Reply:
Most MCUs have plenty timers, it's a pity most board configurations don't expose them (and the timer API in RIOT is mostly catered to software timers, where only one hardware timer is needed. Even though that would probably be better served by SysTick on ARM…)

Reply:
> Maybe the RX implementation could be changed to also periodically sample the input instead. In the timer ISR one would first send the output bit (if any) and then sample the RX bit. The timer would be stopped once both RX and TX are completed. That should both fix the issue with full duplex and only require a single timer.

I had it like that before, well with two timers still.
That would only work if the phases of the RX and the TX signal are aligned though.
Since UART is asynchronous, we can not make such assumptions. We would either have to stretch TX when RX starts or sample the RX bits and random times.

Reply:
> For bringup of a new platform, one would rather not rely on something as fragile as software UART, but rather use e.g. `stdio_semihosting`

This would only work on ARM MCUs. UART is a low hanging fruit. And if timers and GPIO is already implemented, this PR would make it even more low hanging.

When debugging the issue with the UART pins on STM32 going low during initialization, I ended up exposing each initialization step as a shell command in a demo program. That allowed me to not only step throw the initialization, but also modify parameters and the order of initialization; something that On Chip debugging would have not provided in that level of convenience.

Because of the shell, UART is a pretty powerful tool. It would be on my priority list pretty soon after a reliable OpenOCD config and GPIO support.

Reply:
> That would only work if the phases of the RX and the TX signal are aligned though.

I don't think that this is needed, as long as you sample twice the symbol rate. At least one of the two samples should be well within the symbol, no matter what the phase offset is.

I think using a single timer can be the key to reliable full duplex support, as the timing of the ISRs will not interfere with each other. (Still, other IRQs could still break things...)

But I'm not sure if full duplex is actually worth the effort.

Reply:
Unfortunately I don't have much time to work on this currently.
Tbh, I just needed a way to invert the TX signal on an UART and all my hardware UART could do was invert the individual bits. I ended up using the actual hardware UART for RX (because the bus circuitry would invert only the TX signal).

So if you want to pick this and have some better ideas, feel free to go ahead :)

Reply:
Let's call full duplex support not worth the effort for now. The implementation details can still be changed later on.